lib/ansible/modules/cloud/hcloud/hcloud_volume.py

```diff
@@ -0,0 +1,285 @@
+        "labels": {},
+        "server": "my-server",
+    }
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.hcloud import Hcloud
+
+try:
+    from hcloud.volumes.domain import Volume
+    from hcloud.servers.domain import Server
+    import hcloud
+except ImportError:
+    pass
+
+
+class AnsibleHcloudVolume(Hcloud):
+    def __init__(self, module):
+        Hcloud.__init__(self, module, "hcloud_volume")
+        self.hcloud_volume = None
+
+    def _prepare_result(self):
+        server_name = None
+        if self.hcloud_volume.server is not None:
+            server_name = self.hcloud_volume.server.name
+
+        return {
+            "id": to_native(self.hcloud_volume.id),
+            "name": to_native(self.hcloud_volume.name),
+            "size": self.hcloud_volume.size,
+            "location": to_native(self.hcloud_volume.location.name),
+            "labels": self.hcloud_volume.labels,
+            "server": to_native(server_name),
+        }
+
+    def _get_volume(self):
+        try:
+            if self.module.params.get("id") is not None:
+                self.hcloud_volume = self.client.volumes.get_by_id(
+                    self.module.params.get("id")
+                )
+            else:
+                self.hcloud_volume = self.client.volumes.get_by_name(
+                    self.module.params.get("name")
+                )
+        except hcloud.APIException as e:
+            self.module.fail_json(msg=e.message)
+
+    def _create_volume(self):
+        self.module.fail_on_missing_params(
+            required_params=["name", "size"]
+        )
+        params = {
+            "name": self.module.params.get("name"),
+            "size": self.module.params.get("size"),
+            "automount": self.module.params.get("automount"),
+            "format": self.module.params.get("format"),
+            "labels": self.module.params.get("labels")
+        }
+        if self.module.params.get("server") is not None:
+            params['server'] = self.client.servers.get_by_name(self.module.params.get("server"))
+        elif self.module.params.get("location") is not None:
+            params['location'] = self.client.locations.get_by_name(self.module.params.get("location"))
+        else:
+            self.module.warn("server or location is required")
+
+        if not self.module.check_mode:
+            resp = self.client.volumes.create(**params)
+            resp.action.wait_until_finished()
+            [action.wait_until_finished() for action in resp.next_actions]
+
+        self._mark_as_changed()
+        self._get_volume()
+
+    def _update_volume(self):
+        size = self.module.params.get("size")
+        if size:
+            if self.hcloud_volume.size <= size:
+                if not self.module.check_mode:
+                    self.hcloud_volume.resize(size).wait_until_finished()
+                self._mark_as_changed()
```

Comment:
~~~diff
-          if self.hcloud_volume.size <= size:
+          if self.hcloud_volume.size < size:
               if not self.module.check_mode:
                  self.hcloud_volume.resize(size).wait_until_finished()
               self._mark_as_changed()
+          elsif self.hcloud_volume.size > size:
+              self.module.warn("Shrinking of volumes is not supported")

~~~